process-mining/src/com/axonivy/processmining/service/IvyTaskOccurrenceService.java

```diff
@@ -0,0 +1,108 @@
+package com.axonivy.processmining.service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.axonivy.processmining.bo.TaskOccurrence;
+import com.axonivy.processmining.enums.IvyVariable;
+
+import ch.ivyteam.ivy.environment.Ivy;
+import ch.ivyteam.ivy.security.exec.Sudo;
+import ch.ivyteam.ivy.workflow.ITask;
+import ch.ivyteam.ivy.workflow.query.TaskQuery;
+
+public class IvyTaskOccurrenceService {
+	private static final String LIKE_TEXT_SEARCH = "%%%s%%";
+	private static final String SLASH = "/";
+
+	private IvyTaskOccurrenceService() {
+	}
+
+	public static HashMap<String, Integer> countTaskOccurrencesByProcessId(String processId) {
+		HashMap<String, TaskOccurrence> taskOccurrenceMap = getHashMapTaskOccurrencesByProcessId(processId);
+		HashMap<String, Integer> result = correctTaskOccurrences(taskOccurrenceMap);
+
+		// TODO: This block code is used for verification
+		// We will remove it when verification is finished
+		for (Map.Entry<String, Integer> taskOccurrence : result.entrySet()) {
+			Ivy.log().warn("Task {0} occurs {1} times", taskOccurrence.getKey(), taskOccurrence.getValue());
+		}
+
+		return result;
+	}
+
+	private static HashMap<String, TaskOccurrence> getHashMapTaskOccurrencesByProcessId(String processId) {
+		return Sudo.get(() -> {
+			TaskQuery taskQuery = TaskQuery.create().where().requestPath()
+					.isLike(String.format(LIKE_TEXT_SEARCH, processId));
+			HashMap<String, TaskOccurrence> map = new HashMap<>();
+			countTaskOccurrencesByTaskQuery(map, taskQuery);
+			return map;
+		});
+	}
+
+	private static void countTaskOccurrencesByTaskQuery(HashMap<String, TaskOccurrence> map, TaskQuery taskQuery) {
+		List<ITask> tasks = new ArrayList<>();
+		int maxQueryResults = Integer.valueOf(Ivy.var().get(IvyVariable.MAX_QUERY_RESULTS.getVariableName()));
+		Integer startIndex = 0;
+		do {
+			tasks = Ivy.wf().getTaskQueryExecutor().getResults(taskQuery, startIndex, maxQueryResults);
```

Comment:
Instead of we put the query in loop for. Could you try asynchronous functions of java
example code:
```
	private static List<ITask> getAllTaskAccordingToPages(String processId) {
		TaskQuery taskQuery = TaskQuery.create().where().requestPath()
				.isLike(String.format(LIKE_TEXT_SEARCH, processId));
		int maxQueryResults = Integer.parseInt(Ivy.var().get(IvyVariable.MAX_QUERY_RESULTS.getVariableName()));
		long totalPages = Ivy.wf().getTaskQueryExecutor().getResultsPaged(taskQuery, maxQueryResults).pages();

		List<Long> numberPages = new ArrayList<>();
		for (long i = 1; i <= totalPages; i++) {
			numberPages.add(i);
		}

		List<List<Long>> batchPageResult = Lists.partition(numberPages, 5);
		for (List<Long> pages : batchPageResult) {
			List<CompletableFuture<List<ITask>>> completableFutureList = pages.stream()
					.map(pageNumber -> createNewCompleteFutureToQuery(taskQuery, maxQueryResults, pageNumber)).toList();
			List<ITask> allTaskOfPagesAbove = extractListITask(completableFutureList);
		}
		return new ArrayList<>();
	}

	private static CompletableFuture<List<ITask>> createNewCompleteFutureToQuery(TaskQuery taskQuery,
			int maxQueryResults, long pageNumber) {
		return CompletableFuture.supplyAsync(() -> Ivy.wf().getTaskQueryExecutor()
				.getResultsPaged(taskQuery, maxQueryResults).page(Math.toIntExact(pageNumber)));
	}

	private static List<ITask> extractListITask(List<CompletableFuture<List<ITask>>> completableFutureList) {
		completableFutureList.forEach(CompletableFuture::join);
		List<List<ITask>> result = completableFutureList.stream().map(future -> {
			try {
				return future.get();
			} catch (Exception e) {
				throw new RuntimeException(e);
			}
		}).toList();
		return result.stream().flatMap(List::stream).collect(Collectors.toList());
	}
```